### ISSUE TITLE
save_tags function

### ISSUE THREAD

**Issue by timelyportfolio** [rstudio/htmltools#73]
I could have easily missed an easier solution, but saving a `tagList` seems difficult.  Here is the function I wrote to accomplish this using bits and pieces of `htmltools`.  Would this be deemed as something worth of `htmltools`?

```
save_tags <- function (tags, file, selfcontained = F, libdir = "./lib") 
{
    if (is.null(libdir)) {
        libdir <- paste(tools::file_path_sans_ext(basename(file)), 
            "_files", sep = "")
    }
    htmltools::save_html(tags, file = file, libdir = libdir)
    if (selfcontained) {
        if (!htmlwidgets:::pandoc_available()) {
            stop("Saving a widget with selfcontained = TRUE requires pandoc. For details see:\n", 
                "https://github.com/rstudio/rmarkdown/blob/master/PANDOC.md")
        }
        htmlwidgets:::pandoc_self_contained_html(file, file)
        unlink(libdir, recursive = TRUE)
    }
    return(file)
}
```

**Comment by cpsievert** [comment]
IMO, yes, this could have a lot of impact. And ideally, the implementation would not depend in pandoc, but use the base64enc package instead. 

Furthermore, it would be nice if there was an option to just return a character string of HTML (instead of always writing to disk). This would be especially useful for embedding tags/htmlwidgets in Jupyter/nteract/etc notebooks, for instance.

If I can get a 👍  from the maintainer, I may be able to find time for a prototype...

**Comment by cpsievert** [comment]
FWIW, I just implemented something related to this for Jupyter notebooks -- https://github.com/IRkernel/repr/blob/master/R/repr_htmlwidget.r

**Comment by cpsievert** [comment]
Maybe it makes sense to have `renderDependencies()` have a `self_contained` arg (which would base64 encode the dependencies), so other functions, like `renderDocument()` can take advantage

**Comment by cpsievert** [comment]
What we should really do is basically copy https://github.com/ramnathv/htmlwidgets/blob/master/R/pandoc.R over to `{htmltools}` (and use it to implement `htmltools::save_html(selfcontained = TRUE)` which would mean `htmlwidgets::saveWidget()` could be a simple wrapper around `htmltools::save_html()`)

**Comment by cderv** [comment]
Just some thought about this

pandoc.R in htmlwidget seems to be adapted to work with `rmarkdown::pandoc_self_contained_html`
> Forked from htmltools::save_html to work better with pandoc_self_contained_html

However, we need to revisit the function in `{rmarkdown}` https://github.com/rstudio/rmarkdown/issues/2087, and maybe this is something to consider lower level and store in `{pandoc}`

I think we should try synchronize on this Carson when you think you'll tackle this.

**Comment by cpsievert** [comment]
@cderv I'm starting to look into this, and it seems that, unfortunately, htmltools/htmlwidgets may be stuck with using `from = "markdown"` because of https://github.com/ramnathv/htmlwidgets/issues/289 (which originally inspired the fork of `save_html` in https://github.com/ramnathv/htmlwidgets/pull/292). 

Although unfortunate from a managing complexity perspective, I'm not sure htmlwidgets currently suffers from the same class of problems as https://github.com/rstudio/rmarkdown/issues/2087 (because of the recently merged fix in https://github.com/ramnathv/htmlwidgets/pull/401/files)

**Comment by cpsievert** [comment]
@cderv I'm starting to look into this, and it does seem like having essentially the equivalent of 

https://github.com/rstudio/rstudio/blob/0420c87/src/cpp/session/modules/ModuleTools.R#L375

(but with a sensible default for `template`?) exported by `{pandoc}` or `{rmarkdown}` would indeed be quite useful for `{htmltools}` (and thus `{htmlwidgets}`). That said, I may just experiment with a version that's local to `{htmltools}` for now and send you a PR (probably to `{pandoc}`?) if it feels general enough

**Comment by cderv** [comment]
Yes it feels to me general enough. We have https://github.com/rstudio/rmarkdown/issues/2087 for **rmarkdown** function, so we need to rewrite this function. I agree.

**Comment by aloboa** [comment]
Is there a planned date for implementing this feature?
Is there any workaround that could be used in the meanwhile (eg, writing a different format and then use pandoc)?
